File: Advent_of_Code_2023/day1.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <map>

int isNumber(std::string testingString)
{
    std::map<std::string, int> numbersMappings = {
        {"one", 1},
        {"two", 2},
        {"three", 3},
        {"four", 4},
        {"five", 5},
        {"six", 6},
        {"seven", 7},
        {"eight", 8},
        {"nine", 9},
    };

    for (std::pair keyValue : numbersMappings)
    {
        if (testingString.compare(0, keyValue.first.length(), keyValue.first) == 0)
        {
            return keyValue.second;
        }
    }

    return -1;
}

int main(void)
{
    std::ifstream file("inputs/day1.txt");
    std::string line;
    int result_part1 = 0;
    int result_part2 = 0;

    while (std::getline(file, line))
    {
        for (int i = 0; i < line.length(); i++)
        {
            if (isdigit(line[i]))
            {
                result_part1 += 10 * (int(line[i]) - 48);
                result_part2 += 10 * (int(line[i]) - 48);
                break;
            }
            else
            {
                int result = isNumber(line.substr(i));
                if (result > 0)
                {
                    result_part2 += 10 * result;
                    break;
                }
            }
        }

        for (int i = line.length() - 1; i >= 0; i--)
        {
            if (isdigit(line[i]))
            {
                result_part1 += int(line[i]) - 48;
                result_part2 += int(line[i]) - 48;
                break;
            }
            else
            {
                int result = isNumber(line.substr(i));
                if (result > 0)
                {
                    result_part2 += result;
                    break;
                }
            }
        }
    }

    std::cout << result_part1 << std::endl;
    std::cout << result_part2 << std::endl;
}

File: Advent_of_Code_2023/Solutions/day1.py
from Utils.utils import readThisPuzzlesInput

lines = readThisPuzzlesInput(__file__)

def lineStartsWithNumber(line):
    for number in numbersMapping:
        if line.startswith(number):
            return numbersMapping[number]
    return -1

def lineEndsWithNumber(line):
    for number in numbersMapping:
        if line.endswith(number):
            return numbersMapping[number]
    return -1

partOneSum = 0
partTwoSum = 0

numbersMapping = {
    "one" : 1,
    "two" : 2,
    "three" : 3,
    "four" : 4,
    "five" : 5,
    "six" : 6,
    "seven" : 7,
    "eight" : 8,
    "nine" : 9
}

for line in lines:
    firstNum = 0
    lastNum = 0
    for char in line:
        if char.isdigit():
            firstNum = int(char)
            break
    for char in line[::-1]:
        if char.isdigit():
            lastNum = int(char)
            break
    
    partOneSum += firstNum * 10 + lastNum

for line in lines:
    firstNum = 0
    lastNum = 0
    for index in range(0, len(line)):
        char = line[index]
        if char.isdigit():
            firstNum = int(char)
            break
        value = lineStartsWithNumber(line[index:])
        if value != -1:
            firstNum = value
            break
    for index in range(len(line) - 1, -1, -1):
        char = line[index]
        if char.isdigit():
            lastNum = int(char)
            break
        value = lineEndsWithNumber(line[:index + 1])
        if value != -1:
            lastNum = value
            break
    
    partTwoSum += firstNum * 10 + lastNum

print(partOneSum)
print(partTwoSum)